File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/spiders/SpiderQueen.java
package com.nyrds.pixeldungeon.mobs.spiders;

import com.nyrds.pixeldungeon.ai.Hunting;
import com.nyrds.pixeldungeon.game.GameLoop;
import com.nyrds.pixeldungeon.items.chaos.ChaosCrystal;
import com.nyrds.pixeldungeon.items.common.armor.SpiderArmor;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.watabou.pixeldungeon.Badges;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Burning;
import com.watabou.pixeldungeon.actors.buffs.Poison;
import com.watabou.pixeldungeon.actors.mobs.Boss;
import com.watabou.pixeldungeon.items.SpiderCharm;
import com.watabou.pixeldungeon.items.keys.SkeletonKey;
import com.watabou.pixeldungeon.items.weapon.enchantments.Fire;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

public class SpiderQueen extends Boss {
	public SpiderQueen() {
		hp(ht(120));
		baseDefenseSkill = 18;
		baseAttackSkill  = 21;
		expForKill = 11;

		dmgMin = 12;
		dmgMax = 20;
		dr = 10;


		float dice = Random.Float();
		if( dice < 0.33 ) {
			collect(new ChaosCrystal());
		} else if( dice < 0.66 ){
			collect(new SpiderCharm());
		} else{
			collect(new SpiderArmor());
		}

		collect(new SkeletonKey());

		addImmunity( Fire.class );
		addImmunity( Burning.class );
	}

	@Override
    public boolean act(){
		if(Random.Int(0, 20) == 0) {
			CharUtils.spawnOnNextCell(this, "SpiderEgg", (int) (100 * GameLoop.getDifficultyFactor()));
		}
		
		return super.act();
	}
	
	@Override
	public int attackProc(@NotNull Char enemy, int damage ) {
		if (Random.Int( 2 ) == 0) {
			Buff.affect( enemy, Poison.class, Random.Int( 7, 9 ) * Poison.durationFactor( enemy ) );
		}
		
		return damage;
	}
	
	@Override
    public boolean canAttack(@NotNull Char enemy) {
		return adjacent(enemy) && hp() > ht() / 2;
	}
	
	@Override
	public boolean getCloser(int target) {
		if (hp() < ht() / 2) {
			if (getState() instanceof Hunting && level().distance(getPos(), target) < 5) {
				return getFurther(target);
			}
		}
		return super.getCloser(target);
	}

	@Override
	public void die(@NotNull NamedEntityKind cause) {
		Badges.validateBossSlain(Badges.Badge.SPIDER_QUEEN_SLAIN);
		super.die(cause);
	}
}

File: scripts/mobs/SpiderQueen.lua
--[[
  SpiderQueen - spiders lair boss (batch 17c-2a,
  was com.nyrds.pixeldungeon.mobs.spiders.SpiderQueen.java).
  Lays a SpiderEgg about every 21 turns, poisons on hit, below half hp
  she refuses melee and kites away while the target is close. The 1/3
  crystal/charm/armor carry-gear roll rides the spawn hook, one-shot
  guarded (restore re-runs onSpawn). isBoss json carries the die-flow
  and the SkeletonKey.
]]
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

return mob.init{
    act = function(self)
        if math.random(0, 20) == 0 then
            RPD.CharUtils:spawnOnNextCell(self, "SpiderEgg",
                math.floor(100 * RPD.GameLoop:getDifficultyFactor()))
        end
    end,

    attackProc = function(self, enemy, dmg)
        if enemy ~= nil and math.random(2) == 1 then
            RPD.Buffs.Buff:affect(enemy, "Poison",
                math.random(7, 9) * RPD.Buffs.Poison:durationFactor(enemy))
        end
        return dmg
    end,

    canAttack = function(self, enemy)
        return self:hp() > self:ht() / 2 and self:distance(enemy) == 1
    end,

    getCloser = function(self, target, ignorePets)
        if self:hp() < self:ht() / 2
                and self:getState():getTag() == "HUNTING"
                and RPD.Dungeon.level:distance(self:getPos(), target) < 5 then
            return self:getFurther(target)
        end
        return false
    end,

    spawn = function(self, level)
        local data = mob.restoreData(self)
        if data.gearGranted then
            return
        end
        data.gearGranted = true

        local dice = math.random()
        if dice < 0.33 then
            self:collect(RPD.ItemFactory:itemByName("ChaosCrystal"))
        elseif dice < 0.66 then
            self:collect(RPD.ItemFactory:itemByName("SpiderCharm"))
        else
            self:collect(RPD.ItemFactory:itemByName("SpiderArmor"))
        end
    end,

    die = function(self, cause)
        RPD.CharUtils:validateBossSlain("SPIDER_QUEEN_SLAIN")
    end,
}
